src/main/java/org/testng/internal/Invoker.java

```diff
@@ -579,6 +580,29 @@ private ITestResult invokeMethod(
       ITestNGMethod[] afterMethods,
       ConfigurationGroupMethods groupMethods,
       FailureContext failureContext) {
+
+    if(System.getProperties().containsKey("testng.newInstancePerMethod") ||
+       System.getenv().containsKey("TESTNG_NEW_INSTANCE_PER_METHOD") ||
+       instance.getClass().isAnnotationPresent(NewInstancePerMethod.class)) {
+      boolean tryConstructor = true;
+      if (Cloneable.class.isAssignableFrom(instance.getClass())) {
```

Comment:
As pointed out by @cbeust can we instead define an attribute in the new annotation which could capture the factory method's name in the current class. That way if we find this new annotation we can basically look for a method which matches the name provided in the attribute, and just invoke that.